NDNc/pipeline/pipeline-interests-aimd.cpp

```diff
@@ -0,0 +1,186 @@
+        onTimeout();
+
+        if (m_pit->size() >= m_windowSize) {
+            continue;
+        }
+
+        std::vector<PendingInterest> pendingInterests(m_windowSize -
+                                                      m_pit->size());
+        size_t n = m_requestQueue.try_dequeue_bulk(
+            pendingInterests.begin(), m_windowSize - m_pit->size());
+
+        if (n == 0) {
+            continue;
+        }
+
+        std::vector<ndn::Block> interests;
+        interests.reserve(n);
+
+        for (size_t i = 0; i < n; ++i) {
+            auto key = pendingInterests[i].pitEntry;
+
+            m_pit->emplace(std::make_pair(key, pendingInterests[i]));
+            interests.emplace(interests.end(),
+                              std::move(pendingInterests[i].interest));
+
+            m_queue->push(key);
+            m_pit->at(key).markAsExpressed();
+        }
+
+        if (!face->send(std::move(interests))) {
+            std::cout << "FATAL: unable to send Interest packets on face\n";
+            this->end();
+        }
+    }
+}
+
+void PipelineInterestsAimd::onData(std::shared_ptr<ndn::Data> &&data,
+                                   ndn::lp::PitToken &&pitToken) {
+    auto pitEntry = getPITTokenValue(std::move(pitToken));
+
+    if (m_pit->find(pitEntry) == m_pit->end()) {
+        std::cout << "DEBUG: unexpected Data packet dropped\n";
+        return;
+    }
+
+    if (data->getCongestionMark()) {
+        decreaseWindow();
+        std::cout << "Debug: ECN received\n";
+    }
+
+    enqueueData(std::move(data)); // Enqueue Data into Response queue
+    m_pit->erase(pitEntry);
+    increaseWindow();
+}
+
+void PipelineInterestsAimd::onNack(std::shared_ptr<ndn::lp::Nack> &&nack,
+                                   ndn::lp::PitToken &&pitToken) {
+
+    auto pitEntry = getPITTokenValue(std::move(pitToken));
+    std::cout << "INFO: received NACK with reason " << nack->getReason()
+              << "\n";
+
+    switch (nack->getReason()) {
+    case ndn::lp::NackReason::DUPLICATE: {
+        auto interest = getWireDecode(m_pit->at(pitEntry).interest);
+        interest->refreshNonce();
+
+        uint64_t newPITEntry = m_rdn->get();
+        auto lifetime = interest->getInterestLifetime().count();
+
+        auto pendingInterest = PendingInterest(
+            std::move(getWireEncode(std::move(interest), newPITEntry)),
+            newPITEntry, lifetime);
+
+        m_requestQueue.enqueue(std::move(pendingInterest));
+        m_pit->erase(pitEntry);
+        break;
+    }
+    default:
+        enqueueData(nullptr); // Enqueue null to mark error
+        m_pit->erase(pitEntry);
+        break;
+    }
+}
+
+void PipelineInterestsAimd::onTimeout() {
+    while (!m_queue->empty()) {
+        if (m_pit->count(m_queue->front()) == 0) {
+            // Pop entries that were already satisfied with success
+            m_queue->pop();
+            continue;
+        }
+
+        auto entry = m_pit->find(m_queue->front());
```

Comment:
another repeated lookup: the `count` a few lines above and the `find` here. Just do a single `find` and use the returned iterator.